scipy/sparse/compressed.py

```diff
@@ -614,14 +614,32 @@ def __setitem__(self, index, x):
         i, j = self._index_to_arrays(i, j)
 
         if isspmatrix(x):
-            x = x.toarray()
-
-        # Make x and i into the same shape
-        x = np.asarray(x, dtype=self.dtype)
-        x, _ = np.broadcast_arrays(x, i)
+            broadcast_row = x.shape[0] == 1 and i.shape[0] != 1
```

Comment:
Yep, `_index_to_arrays` [makes sure that `i` is two-dimensional](https://github.com/perimosocordiae/scipy/blob/dfcf92e9d094f331c136618009c4c3fd78aff9ab/scipy/sparse/sputils.py#L381). I'm not a huge fan of the current system of converting all indexing into full integer index arrays, but it does make the logic here a lot simpler.